x/rewards/keeper/providers.go

```diff
@@ -0,0 +1,123 @@
+package keeper
+
+import (
+	"cosmossdk.io/math"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/lavanet/lava/utils"
+	"github.com/lavanet/lava/x/rewards/types"
+	subscriptionTypes "github.com/lavanet/lava/x/subscription/types"
+)
+
+func (k Keeper) AggregateRewards(ctx sdk.Context, provider, chainid string, adjustmentDenom uint64, rewards math.Int) {
+	index := types.BasePayIndex{Provider: provider, ChainID: chainid}
+	basepay, found := k.getBasePay(ctx, index)
+	adjustedPay := sdk.NewDecFromInt(rewards).QuoInt64(int64(adjustmentDenom))
+	if !found {
+		basepay = types.BasePay{Total: rewards, TotalAdjusted: adjustedPay}
+	} else {
+		basepay.Total = basepay.Total.Add(rewards)
+		basepay.TotalAdjusted = basepay.TotalAdjusted.Add(adjustedPay)
+	}
+
+	k.setBasePay(ctx, index, basepay)
+}
+
+func (k Keeper) distributeMonthlyBonusRewards(ctx sdk.Context) {
+	total := k.TotalPoolTokens(ctx, types.ProviderRewardsDistributionPool)
+	totalRewarded := sdk.ZeroInt()
+	totalToSendRewarded := sdk.ZeroInt()
+	specs := k.specEmissionParts(ctx)
+	for _, spec := range specs {
+		basepays, totalbasepay := k.specProvidersBasePay(ctx, spec.ChainID)
+		if totalbasepay.IsZero() {
+			continue
+		}
+
+		specTotalPayout := k.specTotalPayout(ctx, total, sdk.NewDecFromInt(totalbasepay), spec)
+		for _, basepay := range basepays {
+			reward := specTotalPayout.Mul(basepay.TotalAdjusted).QuoInt(totalbasepay).TruncateInt()
+			totalRewarded = totalRewarded.Add(reward)
+			if totalRewarded.GT(total) {
+				utils.LavaFormatError("provider rewards are larger than the distribution pool balance", nil,
+					utils.LogAttr("distribution_pool_balance", total.String()),
+					utils.LogAttr("provider_reward", totalRewarded.String()))
+				return
+			}
+			// now give the reward the provider contributor and delegators
+			providerAddr, err := sdk.AccAddressFromBech32(basepay.Provider)
+			if err != nil {
+				continue
+			}
+			_, totalSent, err := k.dualstakingKeeper.RewardProvidersAndDelegators(ctx, providerAddr, basepay.ChainID, reward, string(types.ProviderRewardsDistributionPool), false, false, false)
+			if err != nil {
+				utils.LavaFormatError("Failed to send bonus rewards to provider", err, utils.LogAttr("provider", basepay.Provider))
+			}
+			totalToSendRewarded = totalToSendRewarded.Add(totalSent)
+		}
+	}
+	k.removeAllBasePay(ctx)
+	if !totalToSendRewarded.IsZero() {
+		err := k.bankKeeper.SendCoinsFromModuleToModule(ctx, string(types.ProviderRewardsDistributionPool), subscriptionTypes.ModuleName, sdk.NewCoins(sdk.NewCoin(k.stakingKeeper.BondDenom(ctx), totalToSendRewarded)))
```

Comment:
we don't need this if the change in RewardProvidersAndDelegators sends the tokens while rewarding

Reply:
done